File: kernel/arch/i386/handlers.cpp
#include <kernel/interrupts.hpp>
#include <stdio.h>

void IDT::generic_exception_handler(interrupt_frame*) {
	printf("Unknown error: This shouldn't happen. WTF have I done?\n");
}

void IDT::division_by_zero_handler(interrupt_frame*) {
	printf("Division by zero\n");
}

void IDT::debug_handler(interrupt_frame*) {
	printf("DEBUG\n");
}

void IDT::nmi_handler(interrupt_frame*) {
	printf("Non Maskable Interrupt\n");
}

void IDT::breakpoint_handler(interrupt_frame*) {
	printf("Breakpoint\n");
}

void IDT::overflow_handler(interrupt_frame*) {
	printf("Overfloeeee\n");
}

void IDT::bound_range_exceeded_handler(interrupt_frame*) {
	printf("Bound Range Exceeded\n");
}

void IDT::invalid_opcode_handler(interrupt_frame*) {
	printf("Invalid OPCODE\n");
}

void IDT::device_not_available_handler(interrupt_frame*) {
	printf("Device NOT available\n");
}

void IDT::double_fault_handler(interrupt_frame*) {
	printf("Double fault\n");
}

void IDT::invalid_tss_handler(interrupt_frame*, unsigned int ecode) {
	printf("Invalid TSS\n");
}

void IDT::segment_not_present_handler(interrupt_frame*, unsigned int ecode) {
	printf("Segment not present\n");
}

void IDT::stack_segment_fault_handler(interrupt_frame*, unsigned int ecode) {
	printf("Stack segment fault\n");
}

void IDT::general_protection_fault_handler(interrupt_frame*, unsigned int ecode) {
	printf("General Protection Fault\n");
}

void IDT::page_fault_handler(interrupt_frame*, unsigned int ecode) {
	printf("Page fault\n");
}

void IDT::floating_point_exception_handler(interrupt_frame*) {
	printf("Floating point exception\n");
}

void IDT::alignment_check_handler(interrupt_frame *, unsigned int ecode) {
	printf("Alignment check\n");
}

void IDT::machine_check_handler(interrupt_frame *) {
	printf("Machine check\n");
}

void IDT::simd_floating_point_exception_handler(interrupt_frame*) {
	printf("SIMD floating point exception\n");
}

void IDT::virtualization_exception_handler(interrupt_frame*) {
	printf("VIRT exception\n");
}

void IDT::control_protection_exception_handler(interrupt_frame*, unsigned int ecode) {
	printf("Control Protection Exception\n");
}

void IDT::hypervisor_injection_exception_handler(interrupt_frame*) {
	printf("Hypervisor Injection Exception\n");
}

void IDT::vmm_communication_exception_handler(interrupt_frame*, unsigned int ecode) {
	printf("VMM Communication Exception\n");
}

void IDT::security_exception_handler(interrupt_frame*, unsigned int ecode) {
	printf("Security Exception\n");
}

File: src/kernel/interrupts/handlers.zig
const serial = @import("../serial/serial.zig");
const main = @import("../main.zig");
const std = @import("std");
const spin = @import("../sync/spinlock.zig");
const schman = @import("../scheduler/manager.zig");
const ipcfn = @import("../ipc/ipcfn.zig");

pub const IdtFrame = extern struct {
	ip: u64,
	cs: u64,
	flags: u64,
	sp: u64,
	ss: u64,
	pub fn format(
			self: IdtFrame,
			comptime fmt: []const u8,
			options: std.fmt.FormatOptions,
			writer: anytype,
		) !void {
			_ = fmt;
			_ = options;
			try writer.print("IdtFrame{{ .ip = 0x{X}, .cs = 0x{X}, .flags = 0x{X}, sp = 0x{X}, .ss = 0x{X} }}", .{
				self.ip,
				self.cs,
				self.flags,
				self.sp,
				self.ss
			});
		}
};

const page_fault_ecode = packed struct {
	present: u1,
	write: u1,
	user: u1,
	reserved_bit_set: u1,
	instruction_fetch: u1,
	protection_key_violation: u1,
	shadow_stack: u1,
	reserved: u57
};

fn get_cr2() usize {
	return asm(
		\\ movq %%cr2, %[out]
		: [out] "=r"(->usize)
		:
		:
	);
}

fn send_msg_to_fault_handler(ecode: usize, param1: usize) void {
	const sch = schman.SchedulerManager.get_scheduler_for_cpu(main.mycpuid());
	const meself = sch.current_process.?;
	if(meself.fault_port == -1) {
		sch.exit(meself);
	} else {
		// TODO: send the msg to the fault handler
		var msg = ipcfn.ipc_msg.ipc_message_t {
			.dest = meself.fault_port,
			.value0 = ecode,
			.value1 = param1,
		};
		const s = ipcfn.ipc_send(&msg);
		if(s != ipcfn.ipc_msg.EOK) {
			sch.exit(meself);
		}
		const r = ipcfn.ipc_recv(&msg);
		if(r != ipcfn.ipc_msg.EOK) {
			sch.exit(meself);
		}
	}
}

pub fn page_fault(frame: *IdtFrame, error_code: page_fault_ecode) callconv(.Interrupt) void {
	const cr2 = get_cr2();
	std.log.err("Page fault: {}\n{}\nat address 0x{x} (IP: {x})", .{error_code, frame, cr2, frame.ip});
	//main.hcf();
	send_msg_to_fault_handler(0x14, cr2);
}

pub fn double_fault(frame: *IdtFrame, error_code: u64) callconv(.Interrupt) void {
	std.log.err("Double fault: {} {}", .{error_code, frame});
	main.hcf();
}
pub fn general_protection_fault(frame: *IdtFrame, error_code: u64) callconv(.Interrupt) void {
	std.log.err("General protection fault: {} {}\n", .{error_code, frame});
	//main.hcf();
	send_msg_to_fault_handler(0xD, error_code);
}

pub fn generic_error_code(frame: *IdtFrame, error_code: u64) callconv(.Interrupt) void {
	std.log.err("Unknown exception: {} {}\n", .{error_code, frame});
	//main.hcf();
	send_msg_to_fault_handler(255, error_code);
}

const exception_table = [_][]const u8 {
	"Division Error",
	"Debug",
	"Non-Maskable Interrupt",
	"Breakpoint",
	"Overflow",
	"Bound Range Exceeded",
	"Invalid Opcode",
	"Device Not Available",
	"Double Fault",
	"Coprocessor Segment Overrun",
	"Invalid TSS",
	"Segment Not Present",
	"Stack-Segment Fault",
	"General Protection Fault",
	"Page Fault",
	"Reserved",
	"x87 Floating-Point Exception",
	"Alignment Check",
	"Machine Check",
	"SIMD Floating-Point Exception",
	"Virtualization Exception",
	"Control Protection Exception",
	"Reserved",
	"Reserved",
	"Reserved",
	"Reserved",
	"Reserved",
	"Reserved",
	"Hypervisor Injection Exception",
	"VMM Communication Exception",
	"Security Exception",
	"Reserved",
};

pub fn generic_error_generate(n: u32) fn(*IdtFrame) callconv(.Interrupt) void {
	return struct {
		fn handler(frame: *IdtFrame) callconv(.Interrupt) void {
			std.log.err("Unknown exception #{} ({s}): {}\n", .{ n, exception_table[n], frame });
			//main.hcf();
			send_msg_to_fault_handler(n, 0);
		}
	}.handler;
}
